File: samples/python/text_generation/structured_output_generation.py
#!/usr/bin/env python3
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import json
from typing import Literal

from openvino_genai import GenerationConfig, LLMPipeline, StructuredOutputConfig
from pydantic import BaseModel, Field


class Person(BaseModel):
    name: str = Field(pattern=r"^[A-Z][a-z]{1,20}$")
    surname: str = Field(pattern=r"^[A-Z][a-z]{1,20}$")
    age: int
    city: Literal["Dublin", "Dubai", "Munich"]


class Car(BaseModel):
    model: str = Field(pattern=r"^[A-Z][a-z]{1,20} ?[A-Z][a-z]{0,20} ?.?$")
    year: int
    engine_type: Literal["diesel", "petrol", "electric", "hybrid"]


class Transaction(BaseModel):
    id: int = Field(ge=1000, le=10_000_000)
    amount: float
    currency: Literal["EUR", "PLN", "RUB", "AED", "CHF", "GBP", "USD"]


class ItemQuantities(BaseModel):
    person: int = Field(ge=0, le=100)
    car: int = Field(ge=0, le=100)
    transaction: int = Field(ge=0, le=100)


items_map = {"person": Person, "car": Car, "transaction": Transaction}

sys_message = (
    "You generate JSON objects based on the user's request. You can generate JSON objects with different types of objects: person, car, transaction. "
    "If the user requested a different type, the JSON fields should remain zero. "
    "Please note that the words 'individual', 'person', 'people', 'man', 'human', 'woman', 'inhabitant', 'citizen' are synonyms and can be used interchangeably. "
    'E.g. if the user wants 5 houses, then the JSON must be {"person": 0, "car": 0, "transaction": 0}. '
    'If the user wants 3 people and 1 house, then the JSON must be {"person": 3, "car": 0, "transaction": 0}. '
    "Make sure that the JSON contains the numbers that the user requested. If the user asks for specific attributes, like 'surname', 'model', etc., "
    "ignore this information and generate JSON objects with the same fields as in the schema. "
    "Please use double quotes for JSON keys and values. "
)

sys_message_for_items = "Please try to avoid generating the same JSON objects multiple times."


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("model_dir", help="Path to the model directory. It should contain the OpenVINO model files.")
    args = parser.parse_args()

    device = "CPU"  # GPU can be used as well
    pipe = LLMPipeline(args.model_dir, device)

    config = GenerationConfig()
    config.max_new_tokens = 300

    print(
        "This is a smart assistant that generates structured output in JSON format. "
        "You can ask to generate information about a person, car, or bank transaction. "
        'For example, you can ask: "Please generate jsons for 3 persons and 1 transaction."'
    )

    while True:
        try:
            prompt = input("> ")
        except EOFError:
            break
        pipe.start_chat(sys_message)
        config.structured_output_config = StructuredOutputConfig(
            json_schema=json.dumps(ItemQuantities.model_json_schema())
        )
        config.do_sample = False
        res = json.loads(pipe.generate(prompt, config))
        pipe.finish_chat()
        print(f"Generated JSON with item quantities: {res}")

        config.do_sample = True
        config.temperature = 0.8

        pipe.start_chat(sys_message_for_items)
        generate_has_run = False
        for item, quantity in res.items():
            config.structured_output_config = StructuredOutputConfig(
                json_schema=json.dumps(items_map[item].model_json_schema())
            )
            for _ in range(quantity):
                generate_has_run = True
                json_strs = pipe.generate(prompt, config)
                print(json.loads(json_strs))
        pipe.finish_chat()
        if not generate_has_run:
            print("No items generated. Please try again with a different request.")


if "__main__" == __name__:
    main()

File: samples/js/text_generation/structured_output_generation.js
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

import readline from 'readline';
import { z } from 'zod';
import { LLMPipeline, StructuredOutputConfig } from 'openvino-genai-node';

const PersonSchema = z.object({
    name: z.string().regex(/^[A-Z][a-z]{1,20}$/),
    surname: z.string().regex(/^[A-Z][a-z]{1,20}$/),
    age: z.number(),
    city: z.enum(["Dublin", "Dubai", "Munich"]),
});

const CarSchema = z.object({
    model: z.string().regex(/^[A-Z][a-z]{1,20} ?[A-Z][a-z]{0,20} ?.?$/),
    year: z.number(),
    engine_type: z.enum(["diesel", "petrol", "electric", "hybrid"]),
});

const TransactionSchema = z.object({
    id: z.number().gte(1000).lte(10_000_000),
    amount: z.number(),
    currency: z.enum(["EUR", "PLN", "RUB", "AED", "CHF", "GBP", "USD"]),
});

const ItemQuantitiesSchema = z.object({
    person: z.number().gte(0).lte(100),
    car: z.number().gte(0).lte(100),
    transaction: z.number().gte(0).lte(100),
});

const itemsMap = {
    person: PersonSchema,
    car: CarSchema,
    transaction: TransactionSchema,
};

const sysMessage = `You generate JSON objects based on the user's request. You can generate JSON objects with different types of objects: person, car, transaction. 
If the user requested a different type, the JSON fields should remain zero. 
Please note that the words 'individual', 'person', 'people', 'man', 'human', 'woman', 'inhabitant', 'citizen' are synonyms and can be used interchangeably. 
E.g. if the user wants 5 houses, then the JSON must be {"person": 0, "car": 0, "transaction": 0}. 
If the user wants 3 people and 1 house, then the JSON must be {"person": 3, "car": 0, "transaction": 0}. 
Make sure that the JSON contains the numbers that the user requested. If the user asks for specific attributes, like 'surname', 'model', etc., 
ignore this information and generate JSON objects with the same fields as in the schema. 
Please use double quotes for JSON keys and values.`;

const sysMessageForItems = `Please try to avoid generating the same JSON objects multiple times.`;

async function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    const modelDir = process.argv[2];
    if (!modelDir) {
        console.error('Please provide the path to the model directory as the first argument.');
        process.exit(1);
    }

    const device = 'CPU'; // GPU can be used as well
    const pipe = await LLMPipeline(modelDir, device);

    const config = {};
    config.return_decoded_results = true;
    config.max_new_tokens = 300;

    console.log("This is a smart assistant that generates structured output in JSON format. " +
        "You can ask to generate information about a person, car, or bank transaction. " +
        'For example, you can ask: "Please generate jsons for 3 persons and 1 transaction."');

    async function handleInput(prompt) {
        try {
            await pipe.startChat(sysMessage);
            config.structured_output_config = new StructuredOutputConfig({
                json_schema: JSON.stringify(z.toJSONSchema(ItemQuantitiesSchema))
            });
            config.do_sample = false;

            const json_response = await pipe.generate(prompt, config);
            const res = JSON.parse(json_response);
            await pipe.finishChat();
            console.log(`Generated JSON with item quantities: ${JSON.stringify(res)}`);

            config.do_sample = true;
            config.temperature = 0.8;

            await pipe.startChat(sysMessageForItems);
            let generateHasRun = false;

            for (const [item, quantity] of Object.entries(res)) {
                const schema = itemsMap[item];
                if (!schema) continue;
                config.structured_output_config = new StructuredOutputConfig({
                    json_schema: JSON.stringify(z.toJSONSchema(schema))
                });
                for (let i = 0; i < quantity; i++) {
                    generateHasRun = true;
                    const jsonStr = await pipe.generate(prompt, config);
                    console.log(JSON.parse(jsonStr));
                }
            }

            await pipe.finishChat();

            if (!generateHasRun) {
                console.log("No items generated. Please try again with a different request.");
            }
        } catch (error) {
            console.error("An error occurred:", error);
        }

    }

    rl.on('line', handleInput);
}

main();
